src/pybind/mgr/dashboard/frontend/src/app/shared/services/ceph-task-progress.service.ts

```diff
@@ -0,0 +1,51 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable, NgZone } from '@angular/core';
+import { Router } from '@angular/router';
+
+import { BehaviorSubject, Subscription } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CephTaskProgressService {
+  // Observable sources
+  private progressDataSource = new BehaviorSubject(null);
+
+  // Observable streams
+  progressData$ = this.progressDataSource.asObservable();
+
+  polling: number;
+
+  constructor(private http: HttpClient, private router: Router, private ngZone: NgZone) {
+    this.enablePolling();
+  }
+
+  enablePolling() {
+    this.refresh();
+
+    this.ngZone.runOutsideAngular(() => {
```

Comment:
This is general observation: if I'm not wrong this is 99% the same code as this (SummaryService):

https://github.com/ceph/ceph/blob/cbccab9df52772010af9e62bad8351474fb9431c/src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts#L26-L59

So perhaps there's a chance to improve this. Instead of copying-pasting that snippet, why not creating an `ApiService` (I'm sure there's already a service with that name), that basically does this for every back-end endpoint? That should meet these reqs:
- Configurable URL for back-end API endpoint.
- Configurable polling interval (default 5s?).
- Configurable whether Angular or `runOutisdeAngular` is used: I find this a premature optimization effort, as there are other areas (like heavy methods in HTML templates) that have higher impact on the performance. But in case we use it, I'd encapsulate this, rather than spreading its use across components. Besides, running things outside Angular zone is known to cause some issues with race conditions, async, etc.
- I'd ideally run `Observable.timer()` vs. others, as it removes the (IMHO weird) recursive calls.
- Caching: both observables and values.
- Lazy polling: rather than enabling HTTP polling in the constructor (which forces you to identify whether this component is in Login Page or not), just start polling when the consumer Service/Component calls the service.

Reply:
Thank you for detailed tips, Ernesto :) You're right - making a common service for handling similar operations sounds much better. I am quite new to angular so I have to look into this properly before implementing it.

Reply:
Yeah, I know my suggestion here sounds a bit complicate, but technically it shouldn't be THAT complicate. You could just create a new abstract service (let's call it for example `ApiManagerService`), and the following could be a prototype (it lacks some reqs mentioned above):
```ts
@Injectable({                                                                                           
  providedIn: 'root'                                                                                    
})                                                                                                      
export class ApiManagerService<T> {                                                                    
...
  private poller$: Map<RequestMethod, Observable<T>>;                                                        
                                                                                                        
  constructor(private http: HttpClient, private url: string, private interval=5000 ) { }           
                                                                                                          
  get(): Observable<T> {                                                                           
    if (!this.poller$[RequestMethod.Get]) {
      this.poller$[RequestMethod.Get] = timer(0, this.interval).pipe(                                      
      flatMap(() => this.http.get<T>(this.url)),                                    
      shareReplay(1)                                                                                    
    ); 
    return this.poller$[RequestMethod.Get];                                                                       
  }                                                                                                     
} 

class  ApiProgressService extends ApiManagerService<Progress>
...          
```

Reply:
Upon looking and testing further, I found that the commands sent via CephService's send_command (which call mgr module's send_command) are all sent to the monitor. I tried changing the service name to "mgr" but it threw me an error saying that it's not recognized. I found out that there's a function called json_command in ceph_argparse module that uses its send_command function which seems to accept commands to mgr. they look like so:

def json_command(cluster, target=('mon', ''), prefix=None, argdict=None,
                 inbuf=b'', timeout=0, verbose=False):
    """
    Serialize a command and up a JSON command and send it with send_command() above.

def send_command(cluster, target=('mon', ''), cmd=None, inbuf=b'', timeout=0,
                 verbose=False):
   """
    Send a command to a daemon using librados's
    mon_command, osd_command, mgr_command, or pg_command.  Any bulk input data
    comes in inbuf."""

and in send_command, there's :
     elif target[0] == 'mgr':
            ret, outbuf, outs = run_in_thread(
                cluster.mgr_command, cmd, inbuf, timeout=timeout)

The problem is that I am not sure what cluster I should pass. Am I in the right path? Thanks.